modules/launchd/default.nix

```
@@ -0,0 +1,179 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  inherit (pkgs.stdenv.hostPlatform) isDarwin;
+  inherit (lib.generators) toPlist;
+  inherit (hm.dag) entryAfter;
+
+  cfg = config.launchd;
+  labelPrefix = "org.nixos";
+  dstDir = "${config.home.homeDirectory}/Library/LaunchAgents";
+
+  launchdConfig = { config, name, ... }: {
+    options = {
+      enable = mkEnableOption name;
+      config = mkOption {
+        type = types.submodule (import ./launchd.nix);
+        default = { };
+        example = {
+          ProgramArguments = [ "/usr/bin/say" "Good afternoon" ];
+          StartCalendarInterval = {
+            Hour = 12;
+            Minute = 0;
+          };
+        };
+        description = ''
+          Define a launchd job. See <citerefentry>
+          <refentrytitle>launchd.plist</refentrytitle><manvolnum>5</manvolnum>
+          </citerefentry> for details.
+        '';
+      };
+    };
+
+    config = { config.Label = mkDefault "${labelPrefix}.${name}"; };
+  };
+
+  toAgent = config: pkgs.writeText "${config.Label}.plist" (toPlist { } config);
+
+  agentPlists =
+    mapAttrs' (n: v: nameValuePair "${v.config.Label}.plist" (toAgent v.config))
+    (filterAttrs (n: v: v.enable) cfg.agents);
+
+  agentsDrv = pkgs.runCommand "home-manager-agents" {
+    srcs = attrValues agentPlists;
+    dsts = attrNames agentPlists;
+  } ''
+    mkdir -p "$out"
+
+    if [[ -n "$srcs" ]]; then
+      for (( i=0; i < "''${#srcs[@]}"; i+=1 )); do
+        src="''${srcs[i]}"
+        dst="''${dsts[i]}"
+        ln -s "$src" "$out/$dst"
+      done
+    fi
+  '';
+in {
+  meta.maintainers = with maintainers; [ midchildan ];
+
+  options.launchd.agents = mkOption {
+    type = with types; attrsOf (submodule launchdConfig);
+    default = { };
+    description = "Define launchd agents.";
+  };
+
+  config = mkMerge [
+    {
+      assertions = [{
+        assertion = agentPlists != { } -> pkgs.stdenv.isDarwin;
+        message = let names = lib.concatStringsSep ", " (attrNames agentPlists);
+        in "Must use Darwin for modules that require Launchd: " + names;
+      }];
```

I know the systemd module has the opposite logic, but would it be possible to also add an `enable` option to check against (with a default value of `isDarwin`)?

```suggestion
      assertions = [{
        assertion = (cfg.enable && agentPlists != { }) -> isDarwin;
        message = let names = lib.concatStringsSep ", " (attrNames agentPlists);
        in "Must use Darwin for modules that require Launchd: " + names;
      }];
```

This would make it easier to have services defining both systemd services and launchd agents (until a common backend can be created).

---

I wonder if the same should be done for systemd as well, or otherwise each systemd service would need to be hidden behind an `mkIf` for services using both systemd and launchd.

---

Yeah, we should do the same for systemd services (possibly in a different PR).